docs/proposals/20210628-nodepool-ingress-support.md

```diff
@@ -186,45 +239,122 @@ As an end user, I want to balance the traffic into a NodePool with layer-7 load-
-	- It specifies the ingress class to NodePool name so the ingress CR can select the corresponding ingress controller.
-	- It should be granted the authorities to create/update/patch the related resources in kube-system namespace.
-	- It connects with Yurthub to monitor the resources(ingress/service/configmap/secret/endpoint) of its own NodePool.
-
-3). Ingress controller sidecar:
-	- When sidecar is launched, it gets the NodePool id it belongs to, and connects to Yurthub to list/watch the NodePool.
-	- When it detects the NodePool annotation below is changed, it will update externalIPs of the corresponding service:
-	  ```yaml
-            nodepool.openyurt.io/ingress-ips: "xxx"
-	  ```
-	- When it detects the NodePool annotation below is changed, it will update the corresponding configmap of the NodePool,
-	  which will trigger the ingress controller to reload nginx config for the resource state sync.
-	  ```yaml
-            nodepool.openyurt.io/ingress-config: "xxx"
-	  ```
+#### 1). Yurt-app-manager
+
+Implement a standalone `YurtIngress` operator to support the NodePool ingress feature, which means that `YurtIngress`
+together with `NodePool/YurtAppSet(UnitedDeployment)/YurtAppDaemon` will compose the the current OpenYurt unitization
+capability, and they are all managed by Yurt-app-manager. The responsibility of `YurtIngress` is to monitor the update
+of `YurtIngress` CR and trigger the corresponding operations to nginx ingress controller related components.
+
+`YurtIngress` CRD definition:
+
+```go
+// IngressPool defines the details of a Pool for ingress
+type IngressPool struct {
+	// Indicates the pool name.
+	Name string `json:"name"`
+}
+
+// YurtIngressSpec defines the desired state of YurtIngress
+type YurtIngressSpec struct {
+	// Indicates the replicas of the ingress controllers to be deployed under all the specified nodepools.
+	// +optional
+	Replicas int32 `json:"ingress_controller_replicas_per_pool,omitempty"`
+
+	// Indicates all the nodepools on which to enable ingress.
+	// +optional
+	Pools []IngressPool `json:"pools,omitempty"`
+}
+
+// YurtIngressCondition describes current state of a YurtIngress
+type YurtIngressCondition struct {
+	// Indicates the pools that ingress controller is deployed successfully.
+	IngressReadyPools []string `json:"readypools,omitempty"`
+
+	// Indicates the pools that ingress controller is being deployed or deployed failed.
+	IngressNotReadyPools []map[string]string `json:"unreadypools,omitempty"`
+}
+
+// YurtIngressStatus defines the observed state of YurtIngress
+type YurtIngressStatus struct {
+	// Indicates the number of the ingress controllers deployed under all the specified nodepools.
+	// +optional
+	Replicas int32 `json:"ingress_controller_replicas_per_pool,omitempty"`
+
+	// Indicates all the nodepools conditions on which to enable ingress.
+	// +optional
+	Conditions YurtIngressCondition `json:"conditions,omitempty"`
```

Comment:
how about define Conditions as following:
```
Conditions map[string]YurtIngressCondition

type YurtIngressCondition struct {
	Type YurtIngressConditionType
	Status corev1.ConditionStatus
	LastTransitionTime metav1.Time
	Reason string
	Message string
}

type YurtIngressConditionType string

const (
    IngressPending YurtIngressConditionType = "Pending"
    IngressReady YurtIngressConditionType = "Ready"
    IngressFailure YurtIngressConditionType = "Failure"
    IngressUpdated YurtIngressConditionType = "Updated"
    ...
)

and unitedDeploymentCondition reference is here: 
https://github.com/openyurtio/yurt-app-manager-api/blob/main/pkg/yurtappmanager/apis/apps/v1alpha1/uniteddeployment_types.go
```

Reply:
Okay, I will check and polish this part, then update another version later, thanks!